python/cog/server/http.py

```diff
@@ -86,213 +86,233 @@ def create_app(
-        """
-        Cancel a running prediction
-        """
-        if not runner.is_busy():
-            return JSONResponse({}, status_code=404)
-        try:
-            runner.cancel(prediction_id)
-        except UnknownPredictionError:
-            return JSONResponse({}, status_code=404)
-        else:
+            # FIXME: clean up output files
+            encoded_response = jsonable_encoder(response_object)
+            return JSONResponse(content=encoded_response)
+
+        @app.post("/predictions/{prediction_id}/cancel")
+        async def cancel(prediction_id: str = Path(..., title="Prediction ID")) -> Any:
+            """
+            Cancel a running prediction
+            """
+            if not runner.is_busy():
+                return JSONResponse({}, status_code=404)
+            try:
+                runner.cancel(prediction_id)
+            except UnknownPredictionError:
+                return JSONResponse({}, status_code=404)
+            else:
+                return JSONResponse({}, status_code=200)
+
+        @app.post("/shutdown")
+        async def start_shutdown() -> Any:
+            log.info("shutdown requested via http")
+            if shutdown_event is not None:
+                shutdown_event.set()
             return JSONResponse({}, status_code=200)
 
-    @app.post("/shutdown")
-    async def start_shutdown() -> Any:
-        log.info("shutdown requested via http")
-        if shutdown_event is not None:
-            shutdown_event.set()
-        return JSONResponse({}, status_code=200)
+        async def _check_setup_result() -> Any:
+            if app.state.setup_result is None:
+                return
 
-    async def _check_setup_result() -> Any:
-        if app.state.setup_result is None:
-            return
+            if app.state.health == Health.SETUP_FAILED:
+                # Reset app.state.setup_result so future calls are a no-op
+                app.state.setup_result = None
+                return
 
-        if not app.state.setup_result.done():
-            return
+            if not app.state.setup_result.done():
+                return
 
-        # this can raise CancelledError
-        result = app.state.setup_result.result()
+            # this can raise CancelledError
+            result = app.state.setup_result.result()
 
-        if result.status == schema.Status.SUCCEEDED:
-            app.state.health = Health.READY
-        else:
-            app.state.health = Health.SETUP_FAILED
+            if result.status == schema.Status.SUCCEEDED:
+                app.state.health = Health.READY
+            else:
+                app.state.health = Health.SETUP_FAILED
+
+            app.state.setup_result_payload = result
+
+            # Reset app.state.setup_result so future calls are a no-op
+            app.state.setup_result = None
 
-        app.state.setup_result_payload = result
+    except Exception:
```

Comment:
I think we might want to do a bit of refactoring here so we don't have to wrap 300 lines of the file in `try:/except:` 😆

Reply:
uploaded new iteration.